README.md
# Utility Scripts

The scripts within this project are for useful operations across all supported 
operating systems which include various utilities.

## Table of Contents

* [Warnings](#warnings)
* [Getting Started](#getting-started)
* [Prerequisites](#prerequisites)
* [Setup](#setup)
* [Scripts](#scripts)
	* [BadBlocks](#badblocks)
		* [run_badblocks.sh](#run_badblockssh)
		* [run_badblocks_operational.sh](#run_badblocks_operationalsh)
		* [run_badblocks_destructive.sh](#run_badblocks_destructivesh)
	* [CP](#cp)
		* [copy_all_files_to_dir.sh](#copy_all_files_to_dirsh)
	* [Imaging](#imaging)
		* [run_imaging.sh](#run_imagingsh)
		* [run_imaging_read_write.sh](#run_imaging_read_writesh)
		* [run_imaging_read.sh](#run_imaging_readsh)
		* [run_imaging_write.sh](#run_imaging_writesh)
	* [Processing](#processing)
		* [parallel_commands.sh](#parallel_commandssh)
	* [SCP](#scp)
		* [scp_from_remote.sh](#scp_from_remotesh)
		* [scp_to_remote.sh](#scp_to_remotesh)
	* [SMART](#smart)
		* [read_smart_test_results.sh](#read_smart_test_resultssh)
		* [run_smart_test.sh](#run_smart_testsh)
		* [run_smart_test_with_polling.sh](#run_smart_test_with_pollingsh)
		* [run_smart_conveyance_test.sh](#run_smart_conveyance_testsh)
		* [run_smart_short_test.sh](#run_smart_short_testsh)
		* [run_smart_long_test.sh](#run_smart_long_testsh)
	* [SSH](#ssh)
		* [connect.sh](#connectsh)
		* [run_command.sh](#run_commandsh)
		* [run_script.sh](#run_scriptsh)
	* [Status](#status)
		* [list_drives.sh](#list_drivessh)
		* [list_all_drives.sh](#list_all_drivessh)
		* [list_all_drives_smartctl.sh](#list_all_drives_smartctlsh)
		* [list_all_drives_sysctl.sh](#list_all_drives_sysctlsh)
		* [list_smart_drives.sh](#list_smart_drivessh)
		* [list_smart_drives_smartctl.sh](#list_smart_drives_smartctlsh)
		* [list_smart_drives_sysctl.sh](#list_smart_drives_sysctlsh)
	* [TMUX](#tmux)
		* [start_tmux_session.sh](#start_tmux_sessionsh)
		* [end_tmux_session.sh](#end_tmux_sessionsh)
		* [attach_tmux_session.sh](#attach_tmux_sessionsh)
		* [detach_tmux_session.sh](#detach_tmux_sessionsh)
		* [select_session_pane.sh](#select_session_panesh)
		* [send_tmux_pane_command.sh](#send_tmux_pane_commandsh)
		* [send_tmux_shell_command.sh](#send_tmux_shell_commandsh)
		* [set_layout.sh](#set_layoutsh)
		* [set_layout_tiled.sh](#set_layout_tiledsh)
		* [split_window_horizontal.sh](#split_window_horizontalsh)
		* [split_window_vertical.sh](#split_window_verticalsh)
* [Deployment](#deployment)
* [Dependencies](#dependencies)
* [Notes](#notes)
* [Test Environments](#test-environments)
	* [Operating System Compatibility](#operating-system-compatibility)
	* [Hardware Compatibility](#hardware-compatibility)
* [Contributing](#contributing)
* [Support](#support)
* [Versioning](#versioning)
* [Authors](#authors)
* [Copyright](#copyright)
* [License](#license)
* [Acknowledgments](#acknowledgments)

## Warnings

| :warning: |                      :warning:                       | :warning: |
|   :---:   |                        :---:                         |   :---:   |
| :warning: |   **Executing Imaging scripts may destroy data!**    | :warning: |
| :warning: |                      :warning:                       | :warning: |

## Getting Started

These instructions will get you a copy of the project up and running on your 
local machine for development and testing purposes. See 
[deployment](#deployment) for notes on how to deploy the project on a live 
system.

### Prerequisites

This project does not contain any prerequisites at this time.

### Setup

This project does not contain any setup at this time.  All you need to do is 
place the scripts onto your system and execute the desired functionality.

## Scripts

### BadBlocks

#### run_badblocks.sh

A script to run a BadBlocks test in destructive or operational (Non-Destructive) 
mode.

_Usage_

```
[bash] ./run_badblocks.sh [auto_skip] <drive> [blocksize] [destructive]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|           Drive           |                  Drive to test                   |
|        Block Size         |                Block Size to use                 |
|        Destructive        |               Type of test to run                |

_Commonly Used Block Sizes_

|        Block Size         |                      Value                       |
|           :---:           |                      :---:                       |
|           512b            |                       512                        |
|            1K             |                       1024                       |
|            2K             |                       2048                       |
|            4K             |                       4096                       |
|            8K             |                       8192                       |
|            16K            |                      16384                       |
|            32K            |                      32768                       |
|            64K            |                      65536                       |
|           128K            |                      131072                      |
|           256K            |                      262144                      |
|           512K            |                      524288                      |
|            1M             |                     1048576                      |
|            2M             |                     2097152                      |
|            4M             |                     4194304                      |
|            8M             |                     8388608                      |
|            16M            |                     16777216                     |
|            32M            |                     33554432                     |
|            64M            |                     67108864                     |

_Examples_

* **./run_badblocks.sh** "/dev/da1"
* **./run_badblocks.sh** "/dev/da1" 4096
* **./run_badblocks.sh** "/dev/da1" 4096 true
* **./run_badblocks.sh** "auto_skip" "/dev/da1" 4096 true

_Drives Tested_

|        Status        |                       Component                       |
|        :---:         |                         :---:                         |
|  :white_check_mark:  |                   WD Red Plus (8TB)                   |
|  :white_check_mark:  |                   WD Red Plus (6TB)                   |
|  :white_check_mark:  |                     WD Red (6TB)                      |
|  :white_check_mark:  |                   WD Red Plus (3TB)                   |
|  :white_check_mark:  |                  Kingston 240GB A400                  |

#### run_badblocks_operational.sh

A script to run a BadBlocks test in operational (Non-Destructive) mode.

_Usage_

```
[bash] ./run_badblocks_operational.sh [auto_skip] <drive> [blocksize]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|           Drive           |                  Drive to test                   |
|        Block Size         |                Block Size to use                 |

_Commonly Used Block Sizes_

|        Block Size         |                      Value                       |
|           :---:           |                      :---:                       |
|           512b            |                       512                        |
|            1K             |                       1024                       |
|            2K             |                       2048                       |
|            4K             |                       4096                       |
|            8K             |                       8192                       |
|            16K            |                      16384                       |
|            32K            |                      32768                       |
|            64K            |                      65536                       |
|           128K            |                      131072                      |
|           256K            |                      262144                      |
|           512K            |                      524288                      |
|            1M             |                     1048576                      |
|            2M             |                     2097152                      |
|            4M             |                     4194304                      |
|            8M             |                     8388608                      |
|            16M            |                     16777216                     |
|            32M            |                     33554432                     |
|            64M            |                     67108864                     |

_Examples_

* **./run_badblocks_operational.sh** "/dev/da1"
* **./run_badblocks_operational.sh** "/dev/da1" 4096
* **./run_badblocks_operational.sh** "auto_skip" "/dev/da1" 4096

_Drives Tested_

|        Status        |                       Component                       |
|        :---:         |                         :---:                         |
|  :white_check_mark:  |                   WD Red Plus (8TB)                   |
|  :white_check_mark:  |                   WD Red Plus (6TB)                   |
|  :white_check_mark:  |                     WD Red (6TB)                      |
|  :white_check_mark:  |                   WD Red Plus (3TB)                   |
|  :white_check_mark:  |                  Kingston 240GB A400                  |

#### run_badblocks_destructive.sh

A script to A script to run a BadBlocks test in destructive mode.

_Usage_

```
[bash] ./run_badblocks_destructive.sh [auto_skip] <drive> [blocksize]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|           Drive           |                  Drive to test                   |
|        Block Size         |                Block Size to use                 |

_Commonly Used Block Sizes_

|        Block Size         |                      Value                       |
|           :---:           |                      :---:                       |
|           512b            |                       512                        |
|            1K             |                       1024                       |
|            2K             |                       2048                       |
|            4K             |                       4096                       |
|            8K             |                       8192                       |
|            16K            |                      16384                       |
|            32K            |                      32768                       |
|            64K            |                      65536                       |
|           128K            |                      131072                      |
|           256K            |                      262144                      |
|           512K            |                      524288                      |
|            1M             |                     1048576                      |
|            2M             |                     2097152                      |
|            4M             |                     4194304                      |
|            8M             |                     8388608                      |
|            16M            |                     16777216                     |
|            32M            |                     33554432                     |
|            64M            |                     67108864                     |

_Examples_

* **./run_badblocks_destructive.sh** "/dev/da1"
* **./run_badblocks_destructive.sh** "/dev/da1" 4096
* **./run_badblocks_destructive.sh** "auto_skip" "/dev/da1" 4096

_Drives Tested_

|        Status        |                       Component                       |
|        :---:         |                         :---:                         |
|  :white_check_mark:  |                   WD Red Plus (8TB)                   |
|  :white_check_mark:  |                   WD Red Plus (6TB)                   |
|  :white_check_mark:  |                     WD Red (6TB)                      |
|  :white_check_mark:  |                   WD Red Plus (3TB)                   |
|  :white_check_mark:  |                  Kingston 240GB A400                  |

### CP

#### copy_all_files_to_dir.sh

A script to copy all files recursively into a single directory.

_Usage_

```
[bash] ./copy_all_files_to_dir.sh [auto_skip] <srcDir> <destDir> <filenameStructure> [keepStructure]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|     Source Directory      |                 Source directory                 |
|   Destination Directory   |              Destination directory               |
|    Filename Structure     |             REGEX for file matching              |
|      Keep Structure       |           Keep the directory structure           |

_Examples_

* **./copy_all_files_to_dir.sh** "./Source" "./Destination" "*.txt"
* **./copy_all_files_to_dir.sh** "./Source" "./Destination" "*.txt" false
* **./copy_all_files_to_dir.sh** "auto_skip" "./Source" "./Destination" "*.txt" 
false

### Imaging

#### run_imaging.sh

A script to use the DD command to write data across every bit in a drive or read 
all data bits from a drive or read and write data across every bit in a drive in 
parallel.

_Usage_

```
[bash] ./run_imaging.sh [auto_skip] <drive> [blocksize] [type] [write_random_pattern]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|           Drive           |                  Drive to test                   |
|        Block Size         |                Block Size to use                 |
|           Type            |             The type of test to run              |
|   Write Random Pattern    |              Write a random pattern              |

_Commonly Used Block Sizes_

|        Block Size         |                      Value                       |
|           :---:           |                      :---:                       |
|           512b            |                       512                        |
|            1K             |                       1024                       |
|            2K             |                       2048                       |
|            4K             |                       4096                       |
|            8K             |                       8192                       |
|            16K            |                      16384                       |
|            32K            |                      32768                       |
|            64K            |                      65536                       |
|           128K            |                      131072                      |
|           256K            |                      262144                      |
|           512K            |                      524288                      |
|            1M             |                     1048576                      |
|            2M             |                     2097152                      |
|            4M             |                     4194304                      |
|            8M             |                     8388608                      |
|            16M            |                     16777216                     |
|            32M            |                     33554432                     |
|            64M            |                     67108864                     |

_Types_

|         Type         |                      Description                      |
|        :---:         |                         :---:                         |
|         read         |                Read all bits of drive                 |
|        write         |             Write over all bits of drive              |
|      read-write      |         Read and write over all bits of drive         |

_Examples_

* **./run_imaging.sh** "/dev/da4"
* **./run_imaging.sh** "/dev/da4" 1048576
* **./run_imaging.sh** "/dev/da4" 1048576 "read-write"
* **./run_imaging.sh** "/dev/da4" 1048576 "read-write" true
* **./run_imaging.sh** "auto_skip" "/dev/da4" 1048576 "read-write" true

_Drives Tested_

|        Status        |                       Component                       |
|        :---:         |                         :---:                         |
|  :white_check_mark:  |                   WD Red Plus (8TB)                   |
|  :white_check_mark:  |                   WD Red Plus (6TB)                   |
|  :white_check_mark:  |                     WD Red (6TB)                      |
|  :white_check_mark:  |                   WD Red Plus (3TB)                   |
|  :white_check_mark:  |                  Kingston 240GB A400                  |

#### run_imaging_read_write.sh

A script to use the DD command to read and write data across every bit in a 
drive in parallel.

_Usage_

```
[bash] ./run_imaging_read_write.sh [auto_skip] <drive> [blocksize] [write_random_pattern]
```

_Options_

| Option Flag |                          Description                           |
|    :---:    |                             :---:                              |
|     N/A     |                              N/A                               |

_Parameters_

|         Parameter         |                   Description                    |
|           :---:           |                      :---:                       |
|      Automated Skip       |            Continue without prompting            |
|           Drive           |                  Drive to test                   |
|        Block Size         |                Block Size to use                 |
|   Write Random Pattern    |              Write a random pattern              |

_Commonly Used Block Sizes_

|        Block Size         |                      Value                       |
|           :---:           |                      :---:                       |
|           512b            |                       512                        |
|            1K             |                       1024                       |
|            2K             |                       2048                       |
|            4K             |                       4096                       |
|            8K             |                       8192                       |
|            16K            |                      16384                       |
|            32K            |                      32768                       |
|            64K            |                      65536                       |
|           128K            |                      131072                      |
|           256K            |                      262144                      |
|           512K            |                      524288                      |
|            1M             |                     1048576                      |
|            2M             |                     2097152                      |
|            4M             |                     4194304                      |
|            8M             |                     8388608                      |
|            16M            |                     16777216                     |
|            32M            |                     33554432                     |
|            64M            |                     67108864                     |

_Examples_

* **./run_imaging_read_write.sh** "/dev/da3"
* **./run_imaging_read_write.sh** "/dev/da3" 1048576
* **./run_imaging_read_write.sh** "/dev/da3" 1048576 true
* **./run_imaging_read_write.sh** "auto_skip" "/dev/da3" 1048576 true

_Drives Tested_

|        Status        |                       Component                       |

.github/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) 
and this project adheres to 
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - YYYY-MM-DD

Initial release of the following scripts:

* BadBlocks
	* run_badblocks.sh - A script to run a BadBlocks test in destructive mode or 
	operational mode (Non-Destructive).
	* run_badblocks_destructive.sh - A script to run a BadBlocks test in 
	destructive mode.
	* run_badblocks_operational.sh - A script to run a BadBlocks test in 
	operational mode (Non-Destructive).
* CP (Copy)
	* copy_all_files_to_dir.sh - A script to copy all files recursively into a 
	single directory.
* Imaging
	* run_imaging.sh - A script to use the DD command to write data across 
	every bit in a drive or use the DD command to read all data bits from a 
	drive or use the DD command to write data across every bit in a drive and 
	read all data bits from the same drive in parallel.
	* run_imaging_read.sh - A script to use the DD command to read all data bits 
	from a drive.
	* run_imaging_write.sh - A script to use the DD command to write data across 
	every bit in a drive.
	* run_imaging_read_write.sh - A script to use the DD command to write data 
	across every bit in a drive and read all data bits from the same drive in 
	parallel.
* Processing
	* parallel_commands.sh - A script to execute any number of commands in 
	parallel.
* SCP (Secure Copy)
	* scp_from_remote.sh - A script to securely copy files from a remote server.
	* scp_to_remote.sh - A script to securely copy files to a remote server.
* SMART
	* run_smart_test.sh - A script to run a SMART conveyance, short or long 
	test.
	* run_smart_test_with_polling.sh - A script to run a SMART conveyance, short 
	or long test and continuusly poll for status.
	* run_smart_conveyance_test.sh - A script to run a SMART conveyance test.
	* run_smart_short_test.sh - A script to run a SMART short test.
	* run_smart_long_test.sh - A script to run a SMART long test.
	* read_smart_test_results.sh - A script to run a query for the SMART 
	summary.
* SSH
	* connect.sh - A script to connect to a server using an SSH connection.
	* run_command.sh - A script to execute a command over an SSH connection.
	* run_script.sh - A script to execute a script over an SSH connection.
* Status
	* list_drives.sh - A script to use the SYS or SMART commands to list either 
	all available drives or just those that are SMART-capable drives.
	* list_smart_drives.sh - A script to use the SYS or SMART commands to list 
	available SMART-capable drives.
	* list_smart_drives_smartctl.sh - A script to use the SMART commands to list 
	available SMART-capable drives.
	* list_smart_drives_sysctl.sh - A script to use the SYS CONTROL commands to 
	list available SMART-capable drives.
	* list_all_drives.sh - A script to use the SYS or SMART commands to list all 
	available drives.
	* list_all_drives_smartctl.sh - A script to use the SMART commands to list 
	all available drives.
	* list_all_drives_sysctl.sh - A script to use the SYS CONTROL commands to 
	list all available drives.
* TMUX
	* start_tmux_session.sh - A script to start a new TMUX session.
	* end_tmux_session.sh - A script to end an existing TMUX session.
	* attach_tmux_session.sh - A script to attach to an active TMUX session.
	* detach_tmux_session.sh - A script to detach from an active TMUX session.
	* select_session_pane.sh - A script to select/activate a specific TMUX 
	window pane.
	* send_tmux_pane_command.sh - A script to send and execute a command in a 
	specific TMUX window pane.
	* send_tmux_shell_command.sh - A script to send and execute a shell command 
	in a specific TMUX window.
	* set_layout.sh - A script to set a specific layout for an active TMUX 
	window.
	* set_layout_tiled.sh - A script to set a tiled layout for an active TMUX 
	window.
	* split_window_horizontal.sh - A script to split an active TMUX window 
	horizontally.
	* split_window_vertical.sh - A script to split an active TMUX window 
	vertically.

[//]: # (Version Diffs)
[Unreleased]: https://github.com/jhthorp/Utility-Scripts/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/jhthorp/Utility-Scripts/releases/tag/v1.0.0

imaging/run_imaging.sh
#!/bin/bash
################################################################################
#                           Run A Basic Imaging Test                           #
#                                                                              #
#             This script will export a single function that will              #
#         use the DD command to write data across every bit in a drive         #
#                                      or                                      #
#            use the DD command to read all data bits from a drive             #
#                                      or                                      #
#         use the DD command to write data across every bit in a drive         #
#            and read all data bits from the same drive in parallel            #
################################################################################
#       Copyright © 2020 - 2021, Jack Thorp and associated contributors.       #
#                                                                              #
#    This program is free software: you can redistribute it and/or modify      #
#    it under the terms of the GNU General Public License as published by      #
#    the Free Software Foundation, either version 3 of the License, or         #
#    any later version.                                                        #
#                                                                              #
#    This program is distributed in the hope that it will be useful,           #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             #
#    GNU General Public License for more details.                              #
#                                                                              #
#    You should have received a copy of the GNU General Public License         #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.    #
################################################################################

################################################################################
#                                 SCRIPT SETUP                                 #
################################################################################
#===============================================================================
# This section will store some "Process Global" variables into a stack that
# fully supports nesting into the upcoming includes so that these variables
# are correctly held intact.
#
# The following variables are currently being stored:
#    0 - SOURCING_INVOCATION - Boolean - If the script was sourced not invoked
#    1 - DIR - String - The script's directory path
#===============================================================================
# Get the global stack if it exists
if [ -z ${stack_vars+x} ]; 
then 
  declare stack_vars=(); 
fi

# Determine the BASH source (SOURCING_INVOCATION)
(return 0 2>/dev/null) &&
stack_vars[${#stack_vars[@]}]=1 || 
stack_vars[${#stack_vars[@]}]=0

# Determine the exectuable directory (DIR)
stack_vars[${#stack_vars[@]}]="${BASH_SOURCE%/*}"
if [[ ! -d "${stack_vars[${#stack_vars[@]}-1]}" ]]; 
then 
  stack_vars[${#stack_vars[@]}-1]="${PWD}"; 
fi

################################################################################
#                               SCRIPT INCLUDES                                #
################################################################################
. "${stack_vars[${#stack_vars[@]}-1]}/run_imaging_write.sh"
. "${stack_vars[${#stack_vars[@]}-1]}/run_imaging_read.sh"
. "${stack_vars[${#stack_vars[@]}-1]}/run_imaging_read_write.sh"

################################################################################
#                                  FUNCTIONS                                   #
################################################################################
#===============================================================================
# This function will use the DD command to write data across every bit in a 
# drive or to read all data bits from a drive or to write data across every bit 
# in a drive and read all data bits from the same drive in parallel.
#
# GLOBALS / SIDE EFFECTS:
#   N_A - N/A
#
# OPTIONS:
#   [-na] N/A
#
# ARGUMENTS:
#   [1 - drive] The drive to test
#   [2 - blocksize] The BlockSize to use
#                   Commonly Used Block Sizes:
#                           512b - 512
#                           1K - 1024
#                           2K - 2048
#                           4K - 4096
#                           8K - 8192
#                           16K - 16384
#                           32K - 32768
#                           64K - 65536
#                           128K - 131072
#                           256K - 262144
#                           512K - 524288
#                           1M - 1048576
#                           2M - 2097152
#                           4M - 4194304
#                           8M - 8388608
#                           16M  - 16777216
#                           32M  - 33554432
#                           64M - 67108864
#   [3 - type] The type of test to run
#   [4 - write_random_pattern] Write a random pattern
#
# OUTPUTS:
#   N/A - N/A
#
# RETURN:
#   0 - SUCCESS
#   Non-Zero - ERROR
#===============================================================================
run_imaging ()
{
  declare -r drive=${1}
  declare -r blocksize=${2:-1048576}
  declare -r type=${3:read-write} # Values: read, write, read-write
  declare -r write_random_pattern=${4:-false}

  if [ "${type}" = "write" ]
  then
      run_imaging_write \
        ${drive} \
        ${blocksize} \
        ${write_random_pattern}
  elif [ "${type}" = "read" ]
  then
      run_imaging_read \
        ${drive} \
        ${blocksize}
  else # read-write
      run_imaging_read_write \
        ${drive} \
        ${blocksize} \
        ${write_random_pattern}
  fi
}

################################################################################
#                               SCRIPT EXECUTION                               #
################################################################################
#===============================================================================
# This section will execute if the script is invoked from the terminal rather 
# than sourced into another script as a function.  If the first parameter is 
# "auto_skip" then any prompts will be bypassed.
#
# GLOBALS / SIDE EFFECTS:
#   N_A - N/A
#
# OPTIONS:
#   [-na] N/A
#
# ARGUMENTS:
#   [ALL] All arguments are passed into the script's function except the first 
#         if it is "auto_skip".
#
# OUTPUTS:
#   N/A - N/A
#
# RETURN:
#   0 - SUCCESS
#   Non-Zero - ERROR
#===============================================================================
if [ ${stack_vars[${#stack_vars[@]}-2]} = 0 ]; # SOURCING_INVOCATION
then
  # Print a copyright/license header
  cat << EOF
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| Copyright © 2020 - 2021, Jack Thorp and associated contributors.  |
|          This program comes with ABSOLUTELY NO WARRANTY.          |
|   This is free software, and you are welcome to redistribute it   |
|                     under certain conditions.                     |
|        See the GNU General Public License for more details.       |
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
EOF

  # Collect the type
  if [ "${1}" = "auto_skip" ]
  then
    type=${4:read-write}
  else
    type=${3:read-write}
  fi

  # Prepare the disclaimer
  if [ "${type}" = "read" ]
  then
      read -d '' warning <<- EOF
!!      DEPENDING ON DISK SIZE, THE RUNTIME CAN EXCEED A DAY       !!
EOF
  else # write/read-write
      read -d '' warning <<- EOF
!!        BY PROCEEDING, ALL DATA ON DISK WILL BE DESTROYED        !!
!!       STOP AND DO NOT RUN IF DISK CONTAINS VALUABLE DATA        !!
!!                                                                 !!
!!      DEPENDING ON DISK SIZE, THE RUNTIME CAN EXCEED A DAY       !!
EOF
  fi

  # Print a disclaimer
  cat << EOF

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!                             WARNING                             !!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
${warning}
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
EOF

  if [ "${1}" = "auto_skip" ]
  then
    # Remove the auto_skip parameter
    shift

    # Start the script
    run_imaging "${@}"
  else
    CONTINUE=false
    read -p "Would you like to continue (y/n)?" choice
    case "$choice" in 
      y|Y ) CONTINUE=true;;
      n|N ) CONTINUE=false;;
      * ) echo "Invalid Entry";;
    esac

    if [ "${CONTINUE}" = true ]
    then
        # Start the script
        unset CONTINUE
        run_imaging "${@}"
    fi
    unset CONTINUE
  fi
fi

################################################################################
#                                SCRIPT TEARDOWN                               #
################################################################################
#===============================================================================
# This section will remove the "Process Global" variables from the stack
#===============================================================================
unset stack_vars[${#stack_vars[@]}-1] # DIR
unset stack_vars[${#stack_vars[@]}-1] # SOURCING_INVOCATION